File: app/src/main/java/com/maureen/schedule/CourseViewModel.java
package com.maureen.schedule;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.maureen.schedule.data.AppDatabase;
import com.maureen.schedule.data.CourseInfoBean;

import java.util.List;

/**
 * Function:
 * Create:   2020/11/26
 *
 * @author lianml
 */
public class CourseViewModel extends ViewModel {

    private LiveData<List<CourseInfoBean>> mCourseInfoLiveData;

    public LiveData<List<CourseInfoBean>> getCourseInfoLiveData() {
        if (null == mCourseInfoLiveData) {
            mCourseInfoLiveData = AppDatabase.getInstance().getCourseInfoDao().getAllCourses();
        }
        return mCourseInfoLiveData;
    }

    public CourseInfoBean findCourseById(int courseId) {
        return AppDatabase.getInstance().getCourseInfoDao().getCourseInfoById(courseId);
    }

    public void saveCourseInfo(CourseInfoBean courseInfoBean) {
        AppDatabase.getInstance().getCourseInfoDao().saveCourseInfo(courseInfoBean);
    }

    public void updateCourseInfo(CourseInfoBean courseInfoBean) {
        AppDatabase.getInstance().getCourseInfoDao().updateCourse(courseInfoBean);
    }

    public void deleteCourseInfo(CourseInfoBean courseInfoBean) {
        AppDatabase.getInstance().getCourseInfoDao().delCourse(courseInfoBean);
    }
}

File: app/src/main/java/com/maureen/schedule/CourseViewModel.kt
package com.maureen.schedule

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import com.maureen.schedule.data.AppDatabase
import com.maureen.schedule.data.CourseInfoBean
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Function:
 * Create:   2020/11/26
 *
 * @author lianml
 */
class CourseViewModel : ViewModel() {
    val courseLiveData = MutableLiveData<CourseInfoBean>()

    val courseListLiveData = liveData(Dispatchers.IO) {
        val data = AppDatabase.getInstance().courseInfoDao().getAllCourses()
        emitSource(data)
    }

    fun findCourseById(courseId: Int) = viewModelScope.launch(Dispatchers.IO) {
        val data = AppDatabase.getInstance().courseInfoDao().getCourseInfoById(courseId)
        data?.let {
            courseLiveData.postValue(data!!)
        } ?: run {
            courseLiveData.postValue(CourseInfoBean())
        }

    }

    fun saveCourseInfo(courseInfoBean: CourseInfoBean) = viewModelScope.launch(Dispatchers.IO) {
        AppDatabase.getInstance().courseInfoDao().saveCourseInfo(courseInfoBean)
    }

    fun updateCourseInfo(courseInfoBean: CourseInfoBean) = viewModelScope.launch(Dispatchers.IO) {
        AppDatabase.getInstance().courseInfoDao().updateCourse(courseInfoBean)
    }

    fun deleteCourseInfo(courseInfoBean: CourseInfoBean) = viewModelScope.launch(Dispatchers.IO) {
        AppDatabase.getInstance().courseInfoDao().delCourse(courseInfoBean)
    }
}